feat: UTC-210: Prediction validation

robot-ci-heartex opened the pull request:
<!-- Follow Merge -->
This PR was automatically generated via Follow Merge. [Workflow Run](https://github.com/HumanSignal/label-studio/actions/runs/16887182584)
Please ensure that all linked upstream Pull Requests are merged before proceeding with this one.
- https://github.com/HumanSignal/label-studio-sdk/pull/496
<!-- Follow Merge -->







### Prediction creation fully Covered with Validation

**UI Import (data_import/api.py - ImportPredictionsAPI)**
* Validated: Uses li.validate_prediction() with error collection
* Error Handling: Returns validation errors instead of raising exceptions

**PredictionAPI (tasks/serializers.py - PredictionSerializer)**
* Validated: Has validate() method that uses li.validate_prediction()
* Error Handling: Raises ValidationError with detailed error messages

**Task Bulk Import (tasks/serializers.py - BaseTaskSerializerBulk.add_predictions())**
* Validated: Uses li.validate_prediction() with error collection
* Error Handling: Returns validation errors instead of raising exceptions

**ML Backend Predictions (ml/models.py - MLBackend.predict_tasks())**
* Validated: Uses PredictionSerializer.is_valid() which includes our validation
* Error Handling: Raises ValidationError if validation fails

**Storage Import (io_storages/base_models.py - ImportStorage.add_task())**
* Validated: Uses PredictionSerializer.is_valid() which includes our validation
* Error Handling: Collects all validation errors which will be visible to user under storage with new status "Completed with errors"

Example errors:

**UI import** 
<img width="1748" height="287" alt="image" src="https://github.com/user-attachments/assets/32466926-3594-4c76-a8ec-e265f6895db5" />

**Cloud Sync**
<img width="389" height="391" alt="image" src="https://github.com/user-attachments/assets/2c83a56d-0bda-48e3-bc49-e2e95c949ec6" />
<img width="736" height="641" alt="image" src="https://github.com/user-attachments/assets/30de4f90-05fa-4ab3-8d2e-2339d5601880" />

**Prediction API**
<img width="1015" height="405" alt="image" src="https://github.com/user-attachments/assets/5f88bdce-cdca-4b7a-909f-bfd555515e35" />

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16606979651

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16607124083

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-utc-210'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16607124083

Commit:
prediction validation + tests

Commit:
fix tests with incorrect configs

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16627496691

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-utc-210'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16627496691

Commit:
fix more tests

Commit:
Merge branch 'fb-utc-210' of github.com:heartexlabs/label-studio into fb-utc-210

Commit:
fix moooore tests

Commit:
fix test project setup

Commit:
blue

Commit:
fix test affected by fixture change

Commit:
fix test

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16656486988

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-utc-210'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16656486988

bmartel committed:
fix: UTC-205: Ensure all of feature changes are wrapped in the feature flag (#8074)

Commit:
add test for io_storages prediction validation

Commit:
Merge branch 'fb-utc-210' of github.com:heartexlabs/label-studio into fb-utc-210

bmartel committed:
fix: UTC-205: Ensure that DataManager Annotators CellView handles max display value correctly (#8075)

Commit:
ruff

mcanu committed:
fix: UTC-205: Setting data_manager.max_users_to_display in LSO (#8076)

nass600 committed:
fix: UTC-205: Always return comment authors count (#8078)

Co-authored-by: Marcel Canu <marcel@humansignal.com>

hlomzik committed:
fix: BROS-272: Fix missing autocomplete in Create Project editor (#8072)

Co-authored-by: hlomzik <hlomzik@users.noreply.github.com>

hakan458 commented on label_studio/data_import/functions.py:
```diff
@@ -118,13 +135,108 @@ def set_reimport_background_failure(job, connection, type, value, _):
 
 
-def reformat_predictions(tasks, preannotated_from_fields):
+def reformat_predictions(tasks, preannotated_from_fields, project=None):
```
refactor this to not duplicate validation

Commit:
add pred validation to cloud sync + clean up logs

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16682559232

Commit:
update logic, fix tests

Commit:
Merge branch 'fb-utc-210' of github.com:heartexlabs/label-studio into fb-utc-210

hakan458 committed:
Merge branch 'develop' into fb-utc-210

Commit:
add completed_with_errors status to storages

Commit:
Merge branch 'fb-utc-210' of github.com:heartexlabs/label-studio into fb-utc-210

Commit:
fix test

Commit:
fix error

Commit:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-utc-210

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16733276778

Commit:
fix tests

Commit:
always run validation for import

mcanu commented on label_studio/data_import/api.py:
```diff
@@ -399,22 +423,53 @@ def create(self, request, *args, **kwargs):
+                validation_errors.append(f'Prediction {i}: Error validating prediction - {str(e)}')
+                continue
+
+            # If there are validation errors, raise them before creating any predictions
+            if validation_errors:
+                raise ValidationError(validation_errors)
```
This should go outside the for loop

mcanu commented on label_studio/data_import/functions.py:
```diff
@@ -118,13 +156,106 @@ def set_reimport_background_failure(job, connection, type, value, _):
+
+                if li and li._controls:
+                    # Find a control tag that matches the field name
+                    for control_name, control_tag in li._controls.items():
+                        if control_name == field:
+                            # Use the control's to_name and determine type
+                            if hasattr(control_tag, 'to_name') and control_tag.to_name:
+                                to_name = (
+                                    control_tag.to_name[0]
+                                    if isinstance(control_tag.to_name, list)
+                                    else control_tag.to_name
+                                )
+                                prediction_type = control_tag.tag.lower()
+                            break
```
Can we do  `li.get_control(field)` here?

mcanu commented on label_studio/data_import/functions.py:
```diff
@@ -118,13 +156,106 @@ def set_reimport_background_failure(job, connection, type, value, _):
+                    'model_version': 'preannotated',
+                }
+
+                print(f'\nprediction: {prediction}\n')
```
```suggestion
```

mcanu commented on label_studio/data_import/functions.py:
```diff
@@ -118,13 +156,106 @@ def set_reimport_background_failure(job, connection, type, value, _):
+                print(f'\nprediction: {prediction}\n')
+
+                # Validate the prediction using LabelInterface if project is available
+                # Note: We can't validate here since we don't have project context
```
if `li is not None`, shouldn't we be able to validate?

hakan458 replied:
old comment left here, removing

mcanu commented on label_studio/io_storages/base_models.py:
```diff
@@ -531,9 +598,14 @@ def _scan_and_create_links(self, link_class):
+        if validation_errors:
+            # sync is finished, set completed with errors status for storage info
+            self.info_set_completed_with_errors(
+                last_sync_count=tasks_created, tasks_existed=tasks_existed, validation_errors=validation_errors
```
Where are we showing these validation_errors when Completed with Errors?

hakan458 replied:
In the frontend, same way we view logs for 'Failed' storage syncs 
https://github.com/HumanSignal/label-studio/pull/8062/files#diff-466e193f97b7964093f7265d6b60555d477712bac726e7ce754e7814f38868b2

mcanu commented on label_studio/tasks/serializers.py:
```diff
@@ -395,7 +421,11 @@ def create(self, validated_data):
-            self.add_predictions(task_predictions)
+            prediction_errors = self.add_predictions(task_predictions)
+
+            # If there are prediction validation errors, raise them
+            if prediction_errors:
+                raise ValidationError({'predictions': prediction_errors})
```
This won't call the post_process_* functions after this line.
Do we need to fail before creating the predictions instead? What happens when we raise ValidationError?

hakan458 replied:
I checked and all 3 of those functions do absolutely nothing it seems.... they just `pass` 
Going to keep them there for now in case Im missing something but I think we're good

mcanu replied:
They are probably there to be overriden by LSE. [For example](https://github.com/heartexlabs/label-studio-enterprise/blob/035c679991a095f235a63ab6b46af7a547853d1e/label_studio_enterprise/lse_tasks/serializers.py#L29)

hakan458 committed:
Update label_studio/data_import/functions.py

Co-authored-by: Marcel Canu <marcel@humansignal.com>

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16756473813

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-utc-210'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16756473813

Commit:
clean up

Commit:
Merge branch 'fb-utc-210' of github.com:heartexlabs/label-studio into fb-utc-210

Commit:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-utc-210

Commit:
add test

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16767139146

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-utc-210'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16767139146

niklub committed:
fix: Add beta tag for hotkeys (#8098)

Co-authored-by: nik <nik@heartex.net>
Co-authored-by: niklub <niklub@users.noreply.github.com>

robot-ci-heartex committed:
ci: Update Feature Flags

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16768937385

robot-ci-heartex committed:
ci: Update Feature Flags

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16782351406

dependabot[bot] committed:
chore(deps): bump tmp from 0.2.1 to 0.2.4 in /web (#8107)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

makseq committed:
fix: BROS-290: Fine-tune storage providers rules (#8102)

Commit:
only validate with non-default config

Commit:
Merge branch 'fb-utc-210' of github.com:heartexlabs/label-studio into fb-utc-210

hakan458 commented:
/selenium test

Commit:
update project before calling reimport

Commit:
add taxonomy test

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16821795081

borisheartex commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16825577585)

borisheartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16825577585

borisheartex committed:
Merge branch 'develop' into 'fb-utc-210'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16825577585

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16837060877

hakan458 committed:
Merge branch 'develop' into fb-utc-210

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16887182584

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-utc-210'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16887182584

Commit:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-utc-210

Commit:
Merge branch 'fb-utc-210' of github.com:heartexlabs/label-studio into fb-utc-210

Commit:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-utc-210

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio-enterprise/actions/runs/16922752130
